Fix redis_cache plugin

aero-git opened the pull request:
### What does this PR do?
* Fixed `_build_bank_hier()` function to build a proper hierarchy.
* Fixed `list_()` function to support both types of sets.

### What issues does this PR fix or reference?
Incorrect minion data hierarchy.
Incorrect minion list retrieval.
### Previous Behavior
* When minion data is stored, plugin would add `"."` to `{bank_prefix}{separator}minions` set, instead of `minion_id`.
* `{bank_prefix}{separator}minions/{minion_id}` set is created and `"."` added to it.
* `list_()` function, which is used to fetch the list of minion_ids from cache in `manage.*` runners, would target `{bank_keys_prefix}{separator}minions` set instead of `{bank_prefix}{separator}minions`, returning an empty list.

### New Behavior
* `minion_id` is added to `{bank_prefix}{separator}minions` set.
* `{bank_prefix}{separator}minions/{minion_id}` is not created.
* `list_()` function targets `{bank_prefix}{separator}{bank}` first and if empty list is returned, `{bank_keys_prefix}{separator}{bank}`.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the test documentation for details on how to implement tests
into Salt's test suite:
https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
No

<!-- Please review Salt's Contributing Guide for best practices and guidance in
choosing the right branch:
https://docs.saltproject.io/en/master/topics/development/contributing.html -->

<!-- Additional guidance for pull requests can be found here:
https://docs.saltproject.io/en/master/topics/development/pull_requests.html -->

<!-- See GitHub's page on GPG signing for more information about signing commits
with GPG:
https://help.github.com/articles/signing-commits-using-gpg/ -->

aero-git committed:
Fix redis_cache plugin

* Fixed `_build_bank_hier` function to build a proper hierarchy.
* Fixed `list_ function` to support both types of sets.

dwoz commented:
@aero-git are you willing to write a test for this change?

twangboy reviewed (changes requested):
This needs a changelog as well as a test

Sxderp commented:
With the merge of https://github.com/saltstack/salt/pull/68819, this PR is now obsolete.

dwoz committed:
Add changelog and unit tests for redis_cache fixes

- changelog/67814.fixed.md documents _build_bank_hier and list_ fixes
- tests/pytests/unit/cache/test_redis_cache.py adds regression tests
  covering the dot-sentinel bug in bank hierarchy building and the
  wrong-key-set bug in list_() that caused manage.* to return empty
  minion lists

twangboy commented on salt/cache/redis_cache.py:
```diff
@@ -531,9 +531,12 @@ def list_(bank):
     try:
         banks = redis_server.smembers(bank_redis_key)
+        if not banks:
+            banks = redis_server.smembers(bank_keys_redis_key)
```
In Salt's cache architecture, a bank's list_() method is expected to return both its sub-banks (sub-directories) and its stored data keys (files) simultaneously if both exist.

Using the `if not banks:` fallback means that if a bank happens to have a populated sub-bank, any immediate data keys inside that same bank will be completely hidden from the caller. Furthermore, the return elements need to be explicitly decoded to strings so that downstream processors don't encounter raw bytes.

Suggested Change:
Instead of a conditional fallback, query both sets, combine them using .union(), and cleanly decode the values:

```python
    try:
        banks = redis_server.smembers(bank_redis_key) or set()
        keys = redis_server.smembers(bank_keys_redis_key) or set()
        ret = [val.decode() if isinstance(val, bytes) else val for val in banks.union(keys)]
    except (RedisConnectionError, RedisResponseError) as rerr:
        ...
```

twangboy commented on salt/cache/redis_cache.py:
```diff
@@ -336,14 +336,14 @@ def _build_bank_hier(bank, redis_pipe):
+            redis_pipe.sadd(bank_set, child)
+            parent = f"{parent}/{child}"
+        except:
+            parent = child
```
The hierarchy looping logic should be rewritten using standard Python variables without relying on exception-driven control flows:

```python
    parent = ""
    for child in bank.split("/"):
        if parent:
            bank_set = _get_bank_redis_key(parent)
            parent = f"{parent}/{child}"
        else:
            bank_set = _get_bank_redis_key(child)
            parent = child
            
        log.debug("Adding %s to %s", child, bank_set)
        redis_pipe.sadd(bank_set, child)
```
